lib/plugin-loader.js

```diff
@@ -752,71 +756,275 @@ PluginLoader.prototype = {
+          bootstrapLogger.warn("ZWED0035W", pluginDef.identifier, e);
+          //downstream will pick up error
+          newPlugins.push(Object.assign(pluginDef, {error:e}));
+        }
+      }
+      this.registerStaticPluginsWithManagers(sortedAndRejectedPlugins.plugins);
+      for (const plugin of newPlugins) {
+        this.emit('pluginFound', {
+          data: plugin,
+          count: newPlugins.length
+        });
+      }
+  },
+
+  // Note - Not to be confused with auth capabilities, that describe what an auth plugin can do
+  async getComponentCapabilities(config) {
+    let appServerComp = {};
+    let agentComp = {};
+    let host;
+    let port;
+
+    appServerComp.os = process.platform; // Operating system
+    appServerComp.cpu = process.arch; // CPU architecture
+
+    if (config && config.agent.host) {
+      host = config.agent.host;
+      if (config.agent.http) {
+        port = config.agent.http.port;
+      } else if (config.agent.https) {
+        port = config.agent.https.port;
       }
     }
-    const sortedAndRejectedPlugins = depgraph.processImports();
-    sortedAndRejectedPlugins.plugins = sortedAndRejectedPlugins.plugins.filter((plugin)=> {
-      return !this.pluginMap[plugin.identifier];
+    if (host && port) {
+      let url = 'http://' + host + ':' + port + '/server/agent/environment';
+    
+      return new Promise((resolve, reject) => { /* Obtains and stores environment information from agent */
+        http.get(url, (res) => {
+        const { statusCode } = res; // TODO: Check status code for bad status
+        const contentType = res.headers['content-type'];
+
+        res.setEncoding('utf8');
+        let rawData = '';
+        res.on('data', (chunk) => { rawData += chunk; });
+        res.on('end', () => {
+          try {
+            const parsedData = JSON.parse(rawData);
+            if (parsedData.agentVersion) {
+              agentComp.version = parsedData.agentVersion;
+            }
+            if (parsedData.arch) {
+              agentComp.cpu = parsedData.arch;
+            }
+            if (parsedData.os == undefined && parsedData.hardwareIdentifier && parsedData.osRelease) {
+              agentComp.os = "zos";
+              agentComp.osRelease = parsedData.osRelease;
+              agentComp.hardwareIdentifier = parsedData.hardwareIdentifier;
+            } else {
+              agentComp.os = parsedData.os;
+            }
+            resolve();
+          } catch (e) {
+            bootstrapLogger.severe(e.message);
+            resolve(); // We don't want to reject here. Error gets caught down stream
+          }
+        });
+      }).on('error', (e) => {
+        bootstrapLogger.severe(e.message);
+        resolve();
+      });
+    }).then(() => { /* Obtains and stores the endpoints exposed by the agent */
+
+      'http://' + host + ':' + port + '/server/agent/services';
+
+      return new Promise((resolve, reject) => { 
+        http.get(url, (res) => {
```

Comment:
Please issue only a single HTTP request to obtain services, cache, and re-use the info when needed.

Reply:
This request fires once and then the info gets checked each time per plugin